src/math/src/number.rs

```diff
@@ -1,18 +1,523 @@
+    /// ```
+    /// # use math::Number;
+    ///
+    /// let a = Number::from(0.1);
+    /// let b = Number::from(0.2);
+    ///
+    /// assert_eq!(a.sub(b), Ok(Number::from(-0.1)));
+    /// assert_ne!(b.sub(a), Ok(Number::from(0.1)));
+    /// ```
+    pub fn sub(&self, other: impl Into<Self>) -> Result<Self> {
+        todo!()
+    }
+
+    /// Multiply two numbers
+    ///
+    /// ```
+    /// # use math::Number;
+    ///
+    /// let a = Number::from(0.1);
+    /// let b = Number::from(0.2);
+    ///
+    /// assert_eq!(a.mul(b), Ok(Number::from(-0.1)));
+    /// assert_ne!(b.mul(a), Ok(Number::from(0.1)));
+    /// ```
+    pub fn mul(&self, other: impl Into<Self>) -> Result<Self> {
+        todo!()
+    }
+
+    /// Divide two numbers
+    ///
+    /// # Error
+    /// Return Error::DivisionZero if `other` is 0
+    ///
+    /// ```
+    /// # use math::Number;
+    ///
+    /// # fn main() -> math::Result<()> {
+    ///     let random = Number::random();
+    ///     assert!(random.div(Number::ZERO).is_err());
+    ///     assert_eq!(Number::ZERO.div(random)?, Number::ZERO);
+    ///     assert_eq!(rand.div(rand)?, Number::ONE);
+    ///
+    ///     let a = Number::from(3);
+    ///     let b = Number::from(1.5);
+    ///     let neg_b = Number::from(-1.5);
+    ///     assert_eq!(a.div(b)?, Number::from(2));
+    ///     assert_eq!(a.div(neg_b)?, Number::from(-2));
+    /// #   Ok(())
+    /// # }
+    /// ```
+    pub fn div(&self, other: impl Into<Self>) -> Result<Self> {
+        todo!()
+    }
+
+    /// Raises self to the power of `exp`, using exponentiation by squaring.
+    ///
+    /// ```
+    /// # use math::Number;
+    /// assert_eq!(Number::random().pow(Number::ZERO), Ok(Number::ONE));
+    /// assert_eq!(Number::from(5).pow(2), Ok(Number::from(25)));
+    /// ```
+    pub fn power(&self, exp: impl Into<Self>) -> Result<Self> {
+        todo!()
+    }
+
+    /// Get the remainder of `self / other`
+    ///
+    /// ```
+    /// # use math::Number;
+    /// assert_eq!(Number::from(5).modulo(2), Ok(Number::ONE));
+    /// ```
+    pub fn modulo(&self, other: impl Into<Self>) -> Result<Self> {
+        todo!()
+    }
+
+    /// Get the absolute value of the given number
+    ///
+    /// ```
+    /// # use math::Number;
+    /// let a = Number::random();
+    /// let neg_a = a.mul(-1);
+    ///
+    /// assert_eq!(a.abs(), Ok(a));
+    /// assert_eq!(neg_a.abs(), Ok(a));
+    /// ```
+    pub fn abs(&self) -> Result<Self> {
+        todo!()
+    }
+
+    /// Calculate factorial of a given number
+    /// The number is not limited to integer, it can be a fraction
+    ///
+    /// # Error
+    /// return Error::FactorialNegative if the number is less than 0
+    ///
+    /// ```
+    /// # use math::Number;
+    /// assert_eq!(Number::Zero.factorial(), Number::One);
+    /// assert_eq!(Number::from(5).factorial(), Ok(Number::from(120)));
+    /// assert_eq!(Number::from(3.2).factorial().unwrap().to_string(6), "7.756689");
```

Comment:
`factorial(3.2)` == `7.756689535793181`, which rounded to 6 places is actually `7.756690`

Reply:
Fixed, thanks!